test/general/test_serialization.cpp

```diff
@@ -0,0 +1,349 @@
+TEST(YamlWriter, sharedSpecies)
+{
+    auto original1 = newSolution("ideal-gas.yaml", "simple");
+    auto original2 = newSolution("ideal-gas.yaml", "species-remote");
+
+    YamlWriter writer;
+    writer.addPhase(original1);
+    writer.addPhase(original2);
+    writer.toYamlFile("generated-shared-species.yaml");
+
+    auto duplicate = newSolution("generated-shared-species.yaml", "species-remote");
+    auto thermo1 = original2->thermo();
+    auto thermo2 = duplicate->thermo();
+
+    EXPECT_EQ(thermo1->type(), thermo2->type());
+    EXPECT_EQ(thermo1->speciesNames(), thermo2->speciesNames());
+    EXPECT_DOUBLE_EQ(thermo1->pressure(), thermo2->pressure());
+    EXPECT_DOUBLE_EQ(thermo1->enthalpy_mole(), thermo2->enthalpy_mole());
+}
+
+TEST(YamlWriter, duplicateName)
+{
+    auto original1 = newSolution("ideal-gas.yaml", "simple");
+    auto original2 = newSolution("ideal-gas.yaml", "simple");
+    YamlWriter writer;
+    writer.addPhase(original1);
+    EXPECT_THROW(writer.addPhase(original2), CanteraError);
+}
+
+TEST(YamlWriter, reactions)
+{
+    auto original = newSolution("h2o2.yaml");
+    YamlWriter writer;
+    writer.addPhase(original);
+    writer.setPrecision(14);
+    writer.toYamlFile("generated-h2o2.yaml");
+    auto duplicate = newSolution("generated-h2o2.yaml");
+
+    auto kin1 = original->kinetics();
+    auto kin2 = duplicate->kinetics();
+
+    ASSERT_EQ(kin1->nReactions(), kin2->nReactions());
+    vector_fp kf1(kin1->nReactions()), kf2(kin1->nReactions());
+    kin1->getFwdRateConstants(kf1.data());
+    kin2->getFwdRateConstants(kf2.data());
+    for (size_t i = 0; i < kin1->nReactions(); i++) {
+        EXPECT_NEAR(kf1[i], kf2[i], 1e-13 * kf1[i]) << "for reaction i = " << i;
+    }
+}
+
+TEST(YamlWriter, reaction_units_from_Yaml)
+{
+    auto original = newSolution("h2o2.yaml");
+    YamlWriter writer;
+    writer.addPhase(original);
+    writer.setPrecision(14);
+    writer.setUnits({
+        {"activation-energy", "K"},
+        {"quantity", "mol"},
+        {"length", "cm"}
+    });
+    writer.toYamlFile("generated-h2o2-outunits.yaml");
+    auto duplicate = newSolution("generated-h2o2-outunits.yaml");
+
+    auto kin1 = original->kinetics();
+    auto kin2 = duplicate->kinetics();
+
+    ASSERT_EQ(kin1->nReactions(), kin2->nReactions());
+    vector_fp kf1(kin1->nReactions()), kf2(kin1->nReactions());
+    kin1->getFwdRateConstants(kf1.data());
+    kin2->getFwdRateConstants(kf2.data());
+    for (size_t i = 0; i < kin1->nReactions(); i++) {
+        EXPECT_NEAR(kf1[i], kf2[i], 1e-13 * kf1[i]) << "for reaction i = " << i;
+    }
+}
+
+TEST(YamlWriter, reaction_units_from_Xml)
+{
+    auto original = newSolution("h2o2.xml");
+    YamlWriter writer;
+    writer.addPhase(original);
+    writer.setPrecision(14);
+    writer.setUnits({
+        {"activation-energy", "K"},
+        {"quantity", "mol"},
+        {"length", "cm"}
+    });
+    // Should fail because pre-exponential factors from XML can't be converted
```

Comment:
This is definitely a thorny issue. Because reactions are instantiated from XML without having a `Kinetics` object available, there is no way to determine the dimensions of the rate constant. All you know is that the value itself has already been converted to Cantera's mks+kmol system. I guess you _could_ set the value of `Reaction.rate_units` when calling `Kinetics.addReaction`, which would resolve this for what will presumably be the most common use case, serializing a complete phase definition. But a bare Reaction object created from XML does not contain the information needed to serialize it to a non-default unit system.

Reply:
From what I understand, the reaction order provides most of the missing information, so the main issues are non-standard units as well as volumetric vs surface geometry. Am I seeing this correctly? Regardless, I believe export should work associated kinetics objects are defined (as you mention, this can be done implicitly and is likely the most common use case); this would leave an exception if the reaction isn’t fully set up.

Reply:
Correct, the units are determined by the dimensionality of the various phases, plus the units of the standard concentration for each phase. And while kmol/m^3 or kmol/m^2 are probably the most common, there are several phases where the standard concentration is dimensionless (see the various implmementations of `standardConcentrationUnits`).

Reply:
Thanks for the clarification; I believe we're on the same page now - units do pop up in #995.

Regarding the XML conversion, I still believe that implicitly setting units would make sense.

Reply:
I realized while thinking about this that CTI/XML isn't the only case where you can manage to create a reaction without needing to set it's rate units -- this is also true if you just create a `Reaction` object from scratch. Setting `Reaction.rate_units` now occurs as part of `Kinetics.addReaction`, if it doesn't get set before then. I also ended up refactoring up the `rateCoeffUnits` function and making it a member function of the `Reaction` class (as `calculateRateCoeffUnits`).

Reply:
That’s what had been my suspicion all along. If you look at #995, I have new constructors for `ReactionRate` from `AnyMap` where this will become relevant. Obviously things need to be consolidated as I didn’t have `rate_Units` at my disposal before.